File: Classes/Core Data Categories/ContentAdvisory+TV.m
//
//  ContentAdvisory+TV.m
//  Signal GH
//
//  Created by Glenn Howes on 2/24/14.
//  Copyright (c) 2014 Generally Helpful Software. All rights reserved.
//

#import "ContentAdvisory+TV.h"
#import "ContentAdvisoryDescriptor.h"
#import "ScheduledShow+TV.h"
#import "RatingDescription.h"
#import "LanguageString.h"
#import "EventRating.h"
#import "Rating+TV.h"
#import "LocalizedString+TV.h"

@implementation ContentAdvisory (TV)
+(void) extractAdvisoriesFromDescriptors:(ContentAdvisoryDescriptor*)aTable intoShow:(ScheduledShow*)aShow
{
    NSEntityDescription* contentAdvisory = [NSEntityDescription entityForName:@"ContentAdvisory" inManagedObjectContext:aShow.managedObjectContext];
    
    NSEntityDescription* ratingDescriptionEntity = [NSEntityDescription entityForName:@"RatingDescription" inManagedObjectContext:aShow.managedObjectContext];
    
    NSEntityDescription* eventRatingEntity = [NSEntityDescription entityForName:@"EventRating" inManagedObjectContext:aShow.managedObjectContext];
    for(EventRatingRegion* aRegion in aTable.rating_regions)
    {
        ContentAdvisory* anAdvisory = [[ContentAdvisory alloc] initWithEntity:contentAdvisory insertIntoManagedObjectContext:aShow.managedObjectContext];
        anAdvisory.rating_region = [[NSNumber alloc] initWithInt:aRegion.rating_region];
        NSInteger index = 0;
        for(EventRatingDimension* aDimension in aRegion.eventRatingDimensions)
        {
            EventRating* newEventRating = [[EventRating alloc] initWithEntity:eventRatingEntity insertIntoManagedObjectContext:aShow.managedObjectContext];
            
            
            newEventRating.ratingIndex = [NSNumber numberWithInteger:index++];
            newEventRating.ratingValue = [NSNumber numberWithInteger:aDimension.rating_value];
            
            newEventRating.advisory = anAdvisory;
            [anAdvisory addEventRatingsObject:newEventRating];

        }
        
        for(LanguageString* aLanguageString in aRegion.rating_descriptions)
        {
            RatingDescription* newDescription = [[RatingDescription alloc] initWithEntity:ratingDescriptionEntity insertIntoManagedObjectContext:aShow.managedObjectContext];
            newDescription.locale = aLanguageString.languageCode;
            newDescription.text = aLanguageString.string;
            
            newDescription.advisory = anAdvisory;
            [anAdvisory addRatingDescriptionsObject:newDescription];
        }
        
        anAdvisory.show = aShow;
        [aShow addContentAdvisoriesObject:anAdvisory];
    }
}

+(ContentAdvisory*) retrieveBestMatchFromArrayOfAdvisories:(NSArray*)arrayOfAdvisories
{
    ContentAdvisory* result = nil;
    if(arrayOfAdvisories.count) // TODO actually look to see what region to return
    {
        result = [arrayOfAdvisories lastObject];
    }
    return result;
}

-(NSString*) advisoryStringGivenSetOfRatings:(NSSet*)ratings
{
    NSString* result = nil;
    if(self.ratingDescriptions.count)
    {
        result = [LocalizedString bestMatchFromSet:self.ratingDescriptions];
    }
    
    if(self.eventRatings)
    {
        NSArray* sortedRatings = [ratings.allObjects sortedArrayUsingComparator:^NSComparisonResult(Rating* obj1, Rating* obj2) {
            return [obj1.index compare:obj2.index];
        }];
        for(EventRating* anEventRating in self.eventRatings)
        {
            NSInteger whichDimension = anEventRating.ratingIndex.integerValue;
            if(whichDimension < sortedRatings.count)
            {
                Rating* theRating = sortedRatings[whichDimension];
                NSString* theTitle = [LocalizedString bestMatchFromSet:theRating.titles];
                if(theTitle.length)
                {
                    if(result.length)
                    {
                        result = [result stringByAppendingFormat:@" %@", theTitle];
                    }
                    else
                    {
                        result = theTitle;
                    }
                }
                if(theRating.isGraduated && anEventRating.ratingValue != nil)
                {
                    result = [result stringByAppendingFormat:@" (%@)", anEventRating.ratingValue.stringValue];
                }
            }
        }
    }
    return result;
}

@end

File: Classes/Core Data Categories/ContentAdvisory+TV.swift
//
//  ContentAdvisory+TV.swift
//  Signal GH
//
//  Created by Glenn Howes on 4/7/17.
//  Copyright © 2017 Generally Helpful Software. All rights reserved.
//

import Foundation
import CoreData

public extension ContentAdvisory
{
    public static func extractAdvisories(fromDescriptors advisoryDescriptor: ContentAdvisoryDescriptor, intoShow show: ScheduledShow)
    {
        guard let context = show.managedObjectContext else
        {
            return
        }
        context.performAndWait
        {
            let contentAdvisoryEntity = NSEntityDescription.entity(forEntityName: "ContentAdvisory", in: context)!
            let ratingDescriptionEntity = NSEntityDescription.entity(forEntityName: "RatingDescription", in: context)!
            
            let eventRatingEntity = NSEntityDescription.entity(forEntityName: "EventRating", in: context)!
            
            
            let advisories = advisoryDescriptor.rating_regions.map{ aRegion -> ContentAdvisory in
                
                let anAdvisory = ContentAdvisory(entity: contentAdvisoryEntity, insertInto: context)
                anAdvisory.rating_region = NSNumber(value: aRegion.rating_region)
                
                var index = 0
                let eventRatings = aRegion.eventRatingDimensions.map
                {(aDimension) ->EventRating in
                    let anEventRating = EventRating(entity: eventRatingEntity, insertInto: context)
                    anEventRating.ratingIndex = NSNumber(value: index)
                    index += 1
                    let ratingValue = aDimension.rating_value
                    anEventRating.ratingValue = NSNumber(value: ratingValue)
                    anEventRating.advisory = anAdvisory
                    return anEventRating
                }
                anAdvisory.addToEventRatings(NSSet(array: eventRatings))
                
                
                let descriptions = aRegion.rating_descriptions.map { (aLanguageString) -> RatingDescription in
                    let aDescription = RatingDescription(entity: ratingDescriptionEntity, insertInto: context)
                    aDescription.locale = aLanguageString.languageCode
                    aDescription.text = aLanguageString.string
                    aDescription.advisory = anAdvisory
                    return aDescription
                }
                
                anAdvisory.addToRatingDescriptions(NSSet(array: descriptions))
                
                anAdvisory.show = show
                return anAdvisory
            }
            show.addToContentAdvisories(NSSet(array: advisories))
        }
    }
    public static func retrieveBestMatch(fromAdvisories advisories: Set<ContentAdvisory>) -> ContentAdvisory?
    {
        return advisories.first // TODO, be more selective
    }
    
    public func advisoryString(givenRatings ratings : Set<Rating>) -> String?
    {
        var result : String? = nil
        self.managedObjectContext?.performAndWait {
            if (self.ratingDescriptions?.count)! > 0
            {
                result = LocalizedString.bestMatch(fromSet: self.ratingDescriptions as! Set<LocalizedString>)
            }
            guard let eventRatings = self.eventRatings as? Set<EventRating> else
            {
                return
            }
            let sortedRatings = ratings.sorted(by: { (rating1, rating2) -> Bool in
                guard let index1 = rating1.index?.intValue, let index2 = rating2.index?.intValue else
                {
                    return rating1.index != rating2.index && rating1.index != nil
                }
                return index1 < index2
            })
            
            for anEventRating in eventRatings
            {
                let whichDimension = anEventRating.ratingIndex?.intValue
                if whichDimension ?? 0 < sortedRatings.count
                {
                    let theRating = sortedRatings[whichDimension ?? 0]
                    if let theTitle = LocalizedString.bestMatch(fromSet: theRating.titles as! Set<LocalizedString>)
                    {
                        result = (result != nil) ? "\(result!) \(theTitle)" : theTitle
                    }
                    if theRating.isGraduated?.boolValue ?? false && anEventRating.ratingValue != nil
                    {
                        let ratingValue = anEventRating.ratingValue!.stringValue
                        result = (result != nil) ? "\(result!) ratingValue" : ratingValue
                    }
                }
            }
        }
        return result
    }
}
